### ISSUE TITLE
health_check to support other states than 200

### ISSUE THREAD

**Issue by matti** [kontena/kontena#1790]
currently `health_check:` expects endpoint to return 200.

There are some applications that just don't return 200 from any url (when they are initially deployed, at least)

**Comment by jakolehm** [comment]
Related #1760

**Comment by jnummelin** [comment]
#1760 would change < 400 statuses to be accepted as `healthy´. @matti Would that cover your requirement?

**Comment by matti** [comment]
Yes, but we have other cases where 403 is to be considered healthy (http basic auth means healthy)

**Comment by jnummelin** [comment]
Hmm, that would be problematic since defining http health_check also means that LB will use that same endpoint for checking the health. And HAProxy expects 2xx or 3xx for a healthy check.

**Comment by matti** [comment]
I guess you can't have it all then.

**Comment by jnummelin** [comment]
Well, HAProxy supports also customizing the expected status code, so maybe a regex based approach could be taken.

**Comment by SpComb** [comment]
I think the expected health-check status codes really depend on the application, there is no one correct set of expected http status codes for all applications. Should we support a configurable health check status whitelist instead?

```yaml
health_check:
  protocol: http
  http_status:
  - 200
  - 403
```

This would be some kind of `io.kontena.health_check.http_status=200,403` label, used by the agent `ContainerHealthCheckWorker`, and compiled into a haproxy `http-check expect rstatus ^(200|403)$` config?

**Comment by matti** [comment]
influxdb /ping endpoint returns 204, needs to use /debug/vars that returns 200, but also other stuff

**Comment by SpComb** [comment]
Partially fixed in #3006 for 1.4.1 to accept any 2xx reponse as healthy.

**Comment by SpComb** [comment]
Looks to me like most* other HTTP-based healthchecks are hardcoded/default to accepting 2xx and 3xx as healthy. Maybe we should just tweak the agent http healthcheck to match the behavior of e.g. haproxy, and call it a day so that we can close this issue.

Unclear if we really need/want the extra complexity of configurable healthcheck statuses...
